transport/src/main/java/io/netty/channel/AbstractFileRegion.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright 2017 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel;
+
+import java.io.IOException;
+import java.nio.channels.FileChannel;
+import java.nio.channels.WritableByteChannel;
+
+/**
+ * Base class for {@link FileRegion} implementation.
+ *
+ */
+public abstract class AbstractFileRegion implements FileRegion {
+
+    private long transferIndex;
+
+    /**
+     * Transfers the content of this file region to the specified channel.
+     *
+     * @param target
+     *            the destination of the transfer
+     * @param length
+     *            the maximum number of bytes to transfer
+     */
+    public long transferBytesTo(WritableByteChannel target, long length) throws IOException {
+        long written = transferBytesTo(target, transferIndex, length);
+        if (written > 0) {
+            transferIndex += written;
+        }
+        return written;
+    }
+
+    /**
+     * Returns the {@code transferIndex} of this file region.
+     */
+    public long transferIndex() {
+        return transferIndex;
+    }
+
+    /**
+     * Sets the {@code transferIndex} of file region.
+     *
+     * @throws IndexOutOfBoundsException
+     *             if the specified {@code transferIndex} is less than {@code 0} or greater than
+     *             {@link #count()}
+     */
+    public FileRegion transferIndex(long index) {
+        if (index < 0 || index > count()) {
+            throw new IndexOutOfBoundsException(
+                    String.format("transferIndex: %d (expected: 0 <= transferIndex <= count(%d))",
+                            index, count()));
+        }
+        this.transferIndex = index;
+        return this;
+    }
+
+    /**
+     * Returns {@code true} if and only if {@link #transferableBytes()} is greater than {@code 0}.
+     */
+    public boolean isTransferable() {
+        return transferIndex < count();
```

Comment:
I just follow the implementation of AbstractByteBuf.  `readableBytes` returns `writerIndex - readerIndex` while `isReadable` returns `writerIndex > readerIndex`. I'm OK with using `transferableBytes() > 0`, just want to make sure that there is no other reasons why ByteBuf use `writerIndex > readerIndex` instead of `readableBytes() > 0`